cookbooks/retention.md

```diff
@@ -0,0 +1,214 @@
+        
+```
+
+Now we can load in a subset of `main_summary` and construct the necessary fields for retention calculations:
+
+```python
+ms = spark.sql("""
+    SELECT 
+        client_id, 
+        submission_date_s3,
+        profile_creation_date,
+        os
+    FROM main_summary 
+    WHERE 
+        submission_date_s3 >= '20180401'
+        AND submission_date_s3 <= '20180603'
+        AND sample_id = '42'
+        AND app_name = 'Firefox'
+        AND normalized_channel = 'release'
+        AND os in ('Darwin', 'Windows_NT', 'Linux')
+    """)
+
+PCD_CUTS = ('20180401', '20180415')
+
+ms = (
+    ms.withColumn("pcd", from_unixtime_handler("profile_creation_date")) # i.e. 17500 -> '20171130'
+      .filter("pcd >= '{}'".format(PCD_CUTS[0]))
+      .filter("pcd <= '{}'".format(PCD_CUTS[1]))
+      .withColumn("period", get_period("pcd", "submission_date_s3"))
+)
+```
+
+Note that we filter to profiles that were created in the first half of April so that we have sufficient time to observe 6 weeks of behavior. Now we can calculate retention!
+
+```python
+os_counts = (
+    ms
+    .groupby("os")
+    .agg(F.countDistinct("client_id").alias("total_clients"))
+)
+
+weekly_counts = (
+    ms
+    .groupby("period", "os")
+    .agg(F.countDistinct("client_id").alias("n_week_clients"))
+)
+
+retention_by_os = (
+    weekly_counts
+    .join(os_counts, on='os')
+    .withColumn("retention", F.col("n_clients") / F.col("total_count"))
+)
+```
+
+Peeking at 6-Week Retention
+
+```python
+retention_by_os.filter("period = 6").show()
+```
+
+```
++----------+------+--------------+-------------+-------------------+
+|        os|period|n_week_clients|total_clients|          retention|
++----------+------+--------------+-------------+-------------------+
+|     Linux|     6|          1495|        22422|0.06667558647756668|
+|    Darwin|     6|          1288|         4734|0.27207435572454586|
+|Windows_NT|     6|         29024|       124872|0.23243000832852842|
++----------+------+--------------+-------------+-------------------+
+
+
+```
+
+we observe that 35.6% of Linux users whose profile was created in the first half of April submitted a ping 6 weeks later, and so forth. The example code snippets are consolidated in [this notebook](https://gist.github.com/benmiroglio/fc708e5905fad33b43adb9c90e38ebf4).
+
+
+# New vs. Existing User Retention
+
+The above example calculates **New User Retention**, which is distinct from **Existing User Retention**. This distinction is important when understanding retention baselines (i.e. does this number make sense?). Existing users typically have much higher retention numbers than new users. 
+
+Note that is more common in industry to refer to Existing User Retention as "Churn" (Churn = 1 - Retention), however, we use retention across the board for the sake of consistency and interpretability.
+
+**Please be sure to specify whether or not your retention analysis is for new or existing users.**
+
+
+
```

Comment:
# Note on Point Anchoring vs Period Anchoring

As defined above, N week retention needs some anchor to start N from. 

However, we should note, the anchor itself can be a single point (such as a day) or a period (such as a week), also known as a cohort or acquisition period. 

The code example above is using a point anchor of a day, and defining the N week for each client independently, depending on which day they start. 

Many previous retention analysis at Mozilla, however, uses a period anchor, and calculates retention per cohort. For example: 

* Define a week (20180101 to 20180101) as the anchor/acquisition/cohort period. 
* Define clients who have `profile_creation_date` within these dates as a 'cohort'. 
* Define week Ns as 20180101 + 7N to 20180101 + 7N. 
* For the cohort, calculate each N retention. 
* Repeat for each cohort in the period of interest. 

The major difference between using a 1 week period anchor and a 1 day point anchor is that with the 1 week period anchor, the weekN retention period will be the same for a client who started at the beginning of the anchor period versus a client who started at the end of the anchor period. Whereas for point anchor, those same two clients will have their weekN retention period differ by the amount of days between when they started, respectively. 

**Point anchoring is preferred, however, both methods convey the same information.**